packages/mc-scripts/src/utils/get-config-diff.js

```diff
@@ -0,0 +1,246 @@
+  const submenuLinksDiff = ['submenuLink changed'];
+  const mappedSubmenuLinks = {};
+  oldSubmenuLinks.forEach((oldSubmenuLink) => {
+    mappedSubmenuLinks[oldSubmenuLink.uriPath] = oldSubmenuLink;
+  });
+
+  newSubmenuLinks.forEach((newSubmenuLink) => {
+    const oldSubMenuLink = mappedSubmenuLinks[newSubmenuLink.uriPath];
+    if (newSubmenuLink.uriPath in mappedSubmenuLinks) {
+      const submenuLinkDiff = [
+        `${t(1)}Item with "${newSubmenuLink.uriPath}" was changed`,
+      ];
+      Object.keys(mappedSubmenuLinks[newSubmenuLink.uriPath]).forEach((key) => {
+        if (key === 'defaultLabel') {
+          submenuLinkDiff.push(
+            getStringDiff(
+              oldSubMenuLink.defaultLabel,
+              newSubmenuLink.defaultLabel,
+              'defaultLabel',
+              2
+            )
+          );
+        } else if (key === 'permissions') {
+          const submenuLinkPermissionsDiff = getArrayDiff(
+            oldSubMenuLink.permissions,
+            newSubmenuLink.permissions,
+            3
+          );
+          if (submenuLinkPermissionsDiff.length > 0) {
+            submenuLinkDiff.push(`${t(2)}permission changed`);
+            submenuLinkDiff.push(submenuLinkPermissionsDiff);
+          }
+        } else if (key === 'labelAllLocales') {
+          submenuLinkDiff.push(
+            getLabelAllLocalesDiff(
+              oldSubMenuLink['labelAllLocales'],
+              newSubmenuLink['labelAllLocales'],
+              3
+            )
+          );
+        }
+      });
+      delete mappedSubmenuLinks[newSubmenuLink.uriPath];
+      if (submenuLinkDiff.filter(Boolean).length > 1) {
+        submenuLinksDiff.push(submenuLinkDiff.join('\n'));
+      }
+    } else {
+      submenuLinksDiff.push(
+        `${t(1)}Item with "${green(newSubmenuLink.uriPath)}" was added`
+      );
+    }
+  });
+
+  Object.keys(mappedSubmenuLinks).forEach((key) => {
+    submenuLinksDiff.push(`${t(1)}Item with "${red(key)}" was removed`);
+  });
+
+  if (submenuLinksDiff.length > 1) return submenuLinksDiff.join('\n');
+};
+
+const getConfigDiff = (oldConfig, newConfig) => {
+  const diff = [];
+
+  Object.keys(oldConfig).forEach((key) => {
+    if (notNestedKeySet.has(key)) {
+      diff.push(getStringDiff(oldConfig[key], newConfig[key], key));
+    }
+
+    switch (key) {
+      case 'permissions':
+        diff.push(
+          getPermissionsDiff(oldConfig['permissions'], newConfig['permissions'])
+        );
+        break;
+      case 'mainMenuLink':
+        diff.push(
+          getMainMenuLinkDiff(
+            oldConfig['mainMenuLink'],
+            newConfig['mainMenuLink']
+          )
+        );
+        break;
+      case 'submenuLinks':
+        diff.push(
+          getSubmenuLinksDiff(
+            oldConfig['submenuLinks'],
+            newConfig['submenuLinks']
+          )
+        );
+        break;
+      default:
+        break;
+    }
+  });
```

Comment:
I'm wondering if we can implement this in a more generic and scalable way.

For example, instead of "hardcoding" the fields that we are intend to look for, I would image that we can implement a recursive function that checks e.g. the `typeof` of each field and determines which "diff" function to execute.

This would have the advantage of not having to change this implementation every time we might add/change some of the fields in the config.

Reply:
As discussed, we concluded to leave the hardcoded implementation for now and figure out how we can improve this in a scalable way for when we add/update the fields.